### ISSUE TITLE
Reasons for upcasting the logits dtype outside the kernel

### ISSUE THREAD

**Issue by yzhangcs** [linkedin/Liger-Kernel#241]
Hello, thank you for this great work.
https://github.com/linkedin/Liger-Kernel/blob/acd82728207ebafad28d448640502c108901a967/src/liger_kernel/ops/fused_linear_cross_entropy.py#L69

https://github.com/linkedin/Liger-Kernel/blob/acd82728207ebafad28d448640502c108901a967/src/liger_kernel/ops/fused_linear_cross_entropy.py#L91-L96

I'm wondering if there are any reasons for upcasting/downcasting the logits dtype outside the kernel?
If I understand correctly, we already do fp32 upcast inside, so this op is redundant?
I just compare the outputs of the two versions, i.e., w/ and w/o the upcast, and found there's no precision loss if the above code r removed.

**Comment by ByronHsu** [comment]
Thanks! We did the casting to stay consistent with huggingface behavior. But yes, i think we can do it inside. There is a PR doing this: https://github.com/linkedin/Liger-Kernel/pull/238. 

> found there's no precision loss if the above code r removed.
Curious how did you measure the loss?

Also, very impressive background. Welcome contribution :-)

**Comment by ByronHsu** [comment]
```
 o tensor(0.0005, device='cuda:0', grad_fn=<MaxBackward1>)
```

o difference is a bit high?

**Comment by yzhangcs** [comment]
@ByronHsu Hi, have you compared the final loss of FLCE with the naive counterpart?
I think chunking the input into several pieces might be problematic for very large V and relatively small H. 
For example, if H=1024 and V=128*1024, the `inv_factor` would be 128. Since the accumulated dw are maintained in bf16/fp16, I’m uncertain if this could cause any issues.

It might be better to limit the maximum number of chunks. For instance, we could set `inv_factor` to be `min(8, triton.cdiv(V/H))`.

**Comment by yzhangcs** [comment]
I've also found that for 128K vocab, 8 chunks can be faster, with the cost of nearly <1G additional mem.
This is my benchmarks on H100 using the scripts of https://github.com/sustcsonglin/flash-linear-attention/blob/main/benchmarks/benchmark_training_throughput.py:
| V    | Configuration | TPS (K)   | GiB   |
|:----|:-------------|:-----:|:-----:|
| 32K  | w/ chunk (V/H)     | 46.18 | 35.09 |
|      | w/ chunk (4)      | 46.55 | 35.09 |
|      | w/ chunk (8)      | 46.52 | 35.10 |
|      | w/o chunk     | 46.13 | 39.21 |
| 128K | w/ chunk (V/H)      | 40.80 | 37.66 |
|      | w/ chunk (4)      | 41.82 | 39.03 |
|      | w/ chunk (8)     | 41.80 | 38.05 |
|      | w/o chunk     | 41.12 | 56.05 |

**Comment by yzhangcs** [comment]
UPDATE:

Just trained 3 370M models on 10B tokens of Fineweb-edu with 8K ctx length and 32K vocab
Below are the results. Superisingly, 8 chunks exhibits the best ppl. V/H=32K/1K=32 chunks performs worse than the others.

| | PPL $_\downarrow$ |
|:-------------|:-----:|
| w/ chunk (V/H)     | 12.73  |
| w/ chunk (8)      |  12.70|
| w/o chunk     | 12.71 |

**Comment by ByronHsu** [comment]
Casting was merged into the kernel
